lib/aliquot/payment.rb

```diff
@@ -0,0 +1,135 @@
+module Aliquot
+  ##
+  # A Payment represents a single payment using Google Pay.
+  # It is used to verify/decrypt the supplied token by using the shared secret,
+  # thus avoiding having knowledge of merchant primary keys.
+  class Payment
+    ##
+    # Parameters:
+    # token_string::  Google Pay token (JSON string)
+    # shared_secret:: Base64 encoded shared secret
+    # merchant_id::   Google Pay merchant ID ("merchant:<SOMETHING>")
+    # logger::        The logger to use. Default: Logger.new($stdout)
+    # signing_keys::  Formatted list of signing keys used to sign token contents.
+    #                 Otherwise a thread continuously updating google signing
+    #                 keys will be started.
+    def initialize(token_string, shared_secret, merchant_id, logger: Logger.new($stdout), signing_keys: nil)
+      Aliquot.start_key_updater(logger) if $key_updater_thread.nil? && signing_keys.nil?
```

Comment:
This smells a lot to me:
* `logger` is only sent to the separate thread; why don't we use the logger for other stuff? Perhaps we should accept two loggers? Suggestion: rename named param to `key_updater_logger:` or something alike. I think this should not be something set "per payment", so I think it should go "somewhere else".
* My intuition is that the specs for the threading stuff will be hard to write when this logic is necessary, but I have not dived into it so I cannot explain my point.

Reply:
1. `logger` isn't used as I haven't considered what would be beneficial to log.
2. Well, that depends on what you want to spec :) Generally, I would say spec'ing multithreading correctness is is nigh impossible. We could talk over what you'd want to test.

Reply:
I'd considered creating a factory, but haven't gotten that far :)

Reply:
The main issue stopping this is the question of where such a singleton should be created.

Reply:
1. Ahh, so the logger should likely go; we can easily introduce it if necessary.
2. I'd like as thorough specs as we have for pedicel 🙂

Regarding multi-threading: I totally agree. I'd go quite far to do as little "explicit concurrency code" as possible. IMO it is very hard to get right, and very hard to test.

Could we do with:
* Fetch active certs when process launches.
* If no certs are valid/active/..., re-fetch.

I think I'd even consider leaving it to the client to hand aliquot the certs.